lib/libc/misc/lib_envpath.c

```diff
@@ -0,0 +1,284 @@
+
+ENVPATH_HANDLE envpath_init(FAR const char *name)
+{
+	FAR struct envpath_s *envpath;
+	FAR char *path;
+
+	/* Get the value of the PATH variable */
+
+	path = getenv(name);
+	if (!path) {
+		/* getenv() will return a NULL value if the PATH variable does not
+		 * exist in the environment.
+		 */
+
+		return (ENVPATH_HANDLE) NULL;
+	}
+
+	/* Allocate a container for the PATH variable contents */
+
+	envpath = (FAR struct envpath_s *)
+			  lib_malloc(SIZEOF_ENVPATH_S(strlen(path) + 1));
+
+	if (!envpath) {
+		/* Ooops.. we are out of memory */
+
+		return (ENVPATH_HANDLE) NULL;
+	}
+
+	/* Populate the container */
+
+	strcpy(envpath->path, path);
+	envpath->next = envpath->path;
+
+	/* And return the containing cast to an opaque handle */
+
+	return (ENVPATH_HANDLE) envpath;
+}
+
+/****************************************************************************
+ * Name: envpath_next
+ *
+ * Description:
+ *   Traverse all possible values in the PATH variable in attempt to find
+ *   the full path to an envcutable file when only a relative path is
+ *   provided.
+ *
+ * Input Parameters:
+ *   handle - The handle value returned by envpath_init
+ *   relpath - The relative path to the file to be found.
+ *
+ * Returned Value:
+ *   On success, a non-NULL pointer to a null-terminated string is provided.
+ *   This is the full path to a file that exists in the file system.  This
+ *   function will verify that the file exists (but will not verify that it
+ *   is marked envcutable).
+ *
+ *   NOTE: The string pointer return in the success case points to allocated
+ *   memory.  This memory must be freed by the called by calling lib_free().
+ *
+ *   NULL is returned if no path is found to any file with the provided
+ *   'relpath' from any absolute path in the PATH variable.  In this case,
+ *   there is no point in calling envpath_next() further; envpath_release()
+ *   must be called to release resources set aside by expath_init().
+ *
+ ****************************************************************************/
+
+FAR char *envpath_next(ENVPATH_HANDLE handle, FAR const char *relpath)
+{
+	FAR struct envpath_s *envpath = (FAR struct envpath_s *)handle;
+	struct stat buf;
+	FAR char *endptr;
+	FAR char *path;
+	FAR char *fullpath;
+	int pathlen;
+	int ret;
+
+	/* Verify that a value handle and relative path were provided */
+
+	DEBUGASSERT(envpath && relpath);
+	DEBUGASSERT(relpath[0] != '\0' && relpath[0] != '/');
+
+	/* Loop until (1) we find a file with this relative path from one of the
+	 * absolute paths in the PATH variable, or (2) all of the absolute paths
+	 * in the PATH variable have been considered.
+	 */
+
+	for (;;) {
+		/* Make sure that envpath->next points to the beginning of a string */
+
+		path = envpath->next;
+		if (*path == '\0') {
+			/* If it points to a NULL it means that either (1) the PATH varialbe
+			 * is empty, or (2) we have already examined all of the paths in the
+			 * path variable.
+			 */
+
```

Comment:
add appropriate log before returning NULL, to identify the exits

Reply:
Added, but it might become verbose.